Parallel pull from Share

mitchellwrosen opened the pull request:
![Screenshot from 2022-06-28 14-06-54](https://user-images.githubusercontent.com/1074598/176252500-1ff7c077-cedd-4200-85d0-f4f02e330d30.png)

- [x] Depends on #3155 

## Overview

This PR implements a concurrent pull procedure involving a few different threads:

- One thread that inserts entities fetched from Share into sqlite. Because writing to sqlite requires a lock, we can't do better than that.
- One thread that elaborates subsets of those hashes and adds the elaborated hashes to the work queue (the set of hashes we need to download).
- One thread to coordinate pulling batches of work off of the work queue and spawning one-shot workers, and deciding when there's no more work left to do.

tstat committed:
Report correct number of downloaded entities

mitchellwrosen committed:
begin implementing concurrent pull

mitchellwrosen committed:
implement concurrent pull

mitchellwrosen committed:
break up completeTempEntities a bit so it's easier to understand

mitchellwrosen committed:
more cleanup of concurrent pull code

mitchellwrosen committed:
Add Values literal to SQLite, use it to make elaborateHashes much smarter/faster

tstat committed:
Fix bug in uninsertedHashesVar

mitchellwrosen committed:
update ki hash

mitchellwrosen committed:
stack.yaml.lock changes

mitchellwrosen committed:
⅄ trunk → travis/pull-conc

ChrisPenner commented on unison-cli/src/Unison/Share/Sync.hs:
```diff
@@ -326,112 +330,256 @@ dagbfs goal children =
+
+  workerCount <- newWorkerCount
+
+  Ki.scoped \scope -> do
```
This works really nicely 👍🏼  Good work on this lib

mitchellwrosen replied:
Thanks!

ChrisPenner commented on unison-cli/src/Unison/Share/Sync.hs:
```diff
@@ -326,112 +330,256 @@ dagbfs goal children =
+    elaborator hashesVar uninsertedHashesVar newTempEntitiesQueue workerCount =
+      connect \conn ->
+        forever do
+            (join . atomically) do
```
Is there a benefit to the `(join . atomically) do` pattern here over something like:

```
forever do
  mayNewTempEntities <- atomically $ do ...
  case mayNewTempEntities of
    Nothing -> ...
    Just _ -> ...
```

?

I find interleaving STM and IO makes it tougher to see what's being accomplished transactionally 😅

mitchellwrosen replied:
Sure, we can switch it to not do the `join . atomically` thing. I think instead of a `Maybe (Set Hash)` (or w/e) I'll go with a one-off data type, so it's more clear what going down one branch versus another means.

mitchellwrosen committed:
address PR feedback and do some minor cleanup

also reduce number of downloaders from 10 to 5, because 5 has performed
better in a few tests
